torch_geometric/data/temporal_dataset.py

```diff
@@ -0,0 +1,50 @@
+import copy
+
+import numpy as np
+
+from torch_geometric.data import InMemoryDataset
+
+
+class TemporalDataset(InMemoryDataset):
```

Comment:
I wonder what's the motivation for a `TemporalDataset` class (since it can only hold a single `data` object)? Why can't we simply use `InMemoryDataset` as before?

Wouldn't it make more sense to drop `TemporalDataset` completely and let `TemporalDataLoader` expect `TemporalData` as input rather than `TemporalDataset`?

Reply:
Hi @rusty1s, I've created the `TemporalDataset` class for 2 reasons:
- To get the temporal data ready to be used in a `DataLoader`. Currently, the `DataLoader` expects a `Dataset` or a `List` of `BaseData`. Thus, even passing a list with a single `TemporalData` entry, the mini-batching is not working. I've actually spent some time trying to get it working with only a list of `TemporalData` or a `temporalData` object itself but wasn't able to get it working. It seems the may reason is related to the way that the DataLoader iterates in a `TemporalData` object.
- To abstract the `train_val_test_split` method that can be used in any Temporal Dataset to split it based on the temporal field `t`. This method is actually needed here because we are currently needing to use the dataset with a DataLoader instead of the TemporalData entity directly.

Btw, the `TemporalDataset` as is will also work with the default `DataLoader` implementation (despite being much less performant for this purpose than the `TemporalDataLoader`)

Reply:
I understand. The question I am asking myself is whether we should really include two classes `TemporalData` and `TemporalDataset` into PyG that are nearly identical. IMO, there is no clear difference between these two from a design perspective. Note that, in case `TemporalData` implements `__len__` and `__getitem__` interfaces, it can also be well used as input to `DataLoader`.

The same holds true for `TemporalDataLoader`. Do we actually want people to use `DataLoader` for temporal data when `TemporalDataLoader` exists and is more efficient?

Reply:
> Note that, in case TemporalData implements `__len__` and `__getitem__` interfaces, it can also be well used as input to DataLoader.

Hi @rusty1s, my concern here is the override of the `__len__` code in `TemporalData` (it is currently inheriting that method from `BaseData`) and the not expected collateral effects of that change. The `__len__` method for all `*Data` classes are actually working more similar to a `dictionary` instead of a `list` (it returns the number of keys, not items inside each key). 

> Furthermore, the current design of PyG is that a `Dataset` always holds a list of `data` objects, while `TemporalDataset` breaks with this design choice.

I'm not sure if I understood. The current implementation for `TemporalDataset` is inheriting from `InMemoryDataset` and is holding a `self._data_list` as defined in `InMemoryDataset`. `InMemoryDataset` also has a `self.data` attribute` which is currently used by `TemporalDataset` in `__len__` and `__getitem__`. But it is also used by `InMemoryDataset` indirectly to [get an item](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/data/in_memory_dataset.py#L83).


TLDR: If there is no problem on replacing the `__len__` method of `TemporalData` (and get it working more like a list than a dict) let's do that... :) Otherwise, not sure how to get `DataLoader` working with `TemporalData` without `TemporalDataset`.

Reply:
Thanks for your response @otaviocx. Great discussion! As far as I see, `__getitem__` in `TemporalData` already supports indices, slices, etc (which IMO makes a lot of sense). I think for `__len__`, it is far more reasonable for `len(temporal_data)` to return the number of events rather than the number of keys (I never liked the default of `__len__` on `BaseData` anyway and I don't think it is used anywhere). So, I'm definitely in favor of making `TemporalData` more list-like if you are supportive towards that idea as well. We might want to consider also changing `__iter__` in the sense that we now iterate over the number of events rather than the number of keys.